### PR TITLE
Fix editor crash when unsupported `Resource` is dropped in scene

### LINKED ISSUE #89093: Dragging any Resource not recognized by the editor over the viewport crashes Godot
### Tested versions

v4.3.dev4.official [df78c0636]

### System information

Godot v4.3.dev4 - macOS 14.2.1 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (10 Threads)

### Issue description

When dragging a theme over the viewport, the editor crashes. 

https://github.com/godotengine/godot/assets/2206700/80922fcd-fd60-4706-b4b0-8b0a917360ba

I get this from the crash dialog, not sure if it helps:

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               Godot [94104]
Path:                  /Applications/Godot.app/Contents/MacOS/Godot
Identifier:            org.godotengine.godot
Version:               4.3 (4.3)
Code Type:             ARM-64 (Native)
Parent Process:        launchd [1]
User ID:               501

Date/Time:             2024-03-02 18:35:36.4833 +0100
OS Version:            macOS 14.2.1 (23C71)
Report Version:        12
Anonymous UUID:        33ED798A-8EC3-E728-0035-E4E197B21DF8

Sleep/Wake UUID:       14A2FDC0-AF33-4169-BAFF-3E500503288E

Time Awake Since Boot: 540000 seconds
Time Since Wake:       12305 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGABRT)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000188
Exception Codes:       0x0000000000000001, 0x0000000000000188

Termination Reason:    Namespace SIGNAL, Code 6 Abort trap: 6
Terminating Process:   Godot [94104]

VM Region Info: 0x188 is not in any region.  Bytes before following region: 105554726878840
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      MALLOC_NANO              600060000000-600080000000 [512.0M] rw-/rwx SM=PRV  

Application Specific Information:
abort() called
```

If I don't hover with the dragged resource over the viewport and play a game of moving only around the editor, it doesn't crashes. So I assume that there is something going wrong when trying to make a preview of the resource in the viewport? or something like that 🤷‍♂️

https://github.com/godotengine/godot/assets/2206700/16cd683f-e037-4907-9006-bd232820b74b

### Steps to reproduce

Download the project, drag the `res://UI/MainTheme.tres` into the viewport.

### Minimal reproduction project (MRP)

This is the current project I'm using on the videos. It is not a super minimal version of it, but it is small enough.

[MRP.zip](https://github.com/godotengine/godot/files/14470247/MRP.zip)

### PR DESCRIPTION
**Pull request by luevano**
Follow up on #89106

Change `CanvasItemEditorViewport::can_drop_data` to check if files are valid `Resource` types (`PackedScene` or `Texture2D`) instead of checking for extension types (where `.tres` is valid for `Texture2D` but it's not an exclusive extension for it).

Change safeguard check from #89106 to an error message as suggested. Though, from my testing I can't seem to trigger this error message as the `Resource` type check handles this, too.

Changed nested `if` statements to exit immediately to avoid unnecessary code nesting.

Also fixes #89093, so it actually supersedes #89106

### CODE REVIEW

**Comment by akien-mga** [comment]
CC @ryevdokimov

**Comment by luevano** [comment]
I also fixed an issue that when dropping a scene into itself (cyclic dependency) would create an undo-redo history action and display "Create Node" in the godot diagnostics window; I found it while testing this and it's shown here: https://github.com/godotengine/godot/pull/89106#issuecomment-1975388975. It now behaves like this:

https://github.com/godotengine/godot/assets/55825613/605a582e-39a6-4b51-a4bd-905522885125

The only downside is that there is no easy way to display the "cyclic dependency" error that I know of (that doesn't pollute the logs). Should I include this change in this PR, too?  I think it's related.

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 1]
```diff
@@ -60,6 +60,8 @@
 #include "scene/main/window.h"
 #include "scene/resources/packed_scene.h"
 #include "scene/resources/style_box_texture.h"
+#include <core/error/error_macros.h>
+#include <core/io/resource_loader.h>
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:64]: These are already included indirectly.

**Review by KoBeWi — APPROVED** [review]

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 1 above)
Comment (akien-mga) [reply editor/plugins/canvas_item_editor_plugin.cpp:64]: Also for future PRs, such includes should use `""`, not `<>`. The latter is reserved for system headers.

**Review by ryevdokimov — APPROVED** [review]
Looks good.

https://github.com/godotengine/godot/assets/105675984/3b419cde-6a20-42de-b1d2-7e0b7e25fafe

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 1 above)
Comment (luevano) [reply editor/plugins/canvas_item_editor_plugin.cpp:64]: Sure, didn't notice as they got added by mi text editor; will change.

**Comment by luevano** [comment]
It's a change on the same `can_drop_data` and minor one in the `_cyclical_dependency_exists` to mark it as `const`.

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 1 above)
Comment (luevano) [reply editor/plugins/canvas_item_editor_plugin.cpp]: Rebased and removed includes.

**Comment by KoBeWi** [comment]
I think that other change should be in `_perform_drop_data()`, to prevent creating undo actions if no node was created. Thus it's better as a separate fix.

**Comment by luevano** [comment]
To keep this simple and prioritize the crash fix, I'll leave it as is. Once this is merged I can create a PR for the change I'm describing

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**